In [3]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import  LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
import pandas as pd


In [18]:
df = pd.read_csv("fake_job_postings.csv")

In [ ]:
# columns_needed = [
#     'title',
#     'company_profile',
#     'description',
#     'requirements',
#     'benefits',
#     'employment_type',
#     'required_experience',
#     'required_education',
#     'fraudulent'
# ]
# df = df[columns_needed]

In [20]:
df['fraudulent'].value_counts()

fraudulent
0    17014
1      866
Name: count, dtype: int64

In [23]:
text_columns = ['title','company_profile','description','requirements','benefits']

for col in text_columns:
    df[col]= df[col].fillna('')

for col in text_columns:
    df[col]=df[col].str.lower()


categorical_columns = ['employment_type','required_experience','required_education']

for col in categorical_columns:
    df[col]= df[col].fillna("Unknown")

In [30]:
from sklearn.model_selection import train_test_split

X = df.drop("fraudulent",axis=1)
y = df['fraudulent']

X_train  , X_test , y_train , y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [ ]:
text_columns = ['title','company_profile','description','requirements','benefits']
categorical_columns = ['employment_type','required_experience','required_education']

numerical_columns = []

for col in X_train.columns:
    if X_train[col].dtype in ['int64','float64']:
        numerical_columns.append(col)


int64    4
Name: count, dtype: int64

from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import  LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
import pandas as pd

In [64]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(
    n_estimators=300,
    max_depth=None,
    min_samples_split=5,
    min_samples_leaf=2,
    class_weight="balanced",  # VERY IMPORTANT
    random_state=42,
    n_jobs=-1
)

In [65]:
text_transformer = TfidfVectorizer(max_features=500,stop_words="english")
categorical_transformer = OneHotEncoder(handle_unknown="ignore",sparse_output=False)

preprocessor = ColumnTransformer([
    ('text',text_transformer,'description'),
    ("categorical",categorical_transformer,categorical_columns),
    ('numerical','passthrough',numerical_columns)
])

pipeline = Pipeline([
    ("preprocessor",preprocessor),
    ("model",model)
])

In [49]:
import warnings
warnings.filterwarnings("ignore")


In [66]:
pipeline.fit(X_train,y_train)
y_pred = pipeline.predict(X_test)

In [67]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1_value = f1_score(y_test, y_pred)
Confusion_M = confusion_matrix(y_test, y_pred)

print("Accuracy",accuracy)
print("precision",precision)
print("recall",recall)
print("f1_score",f1_value)
print("confusion",Confusion_M)

Accuracy 0.982662192393736
precision 0.983739837398374
recall 0.6685082872928176
f1_score 0.7960526315789473
confusion [[3393    2]
 [  60  121]]


In [60]:
from sklearn.linear_model import  LogisticRegression

model = LogisticRegression(max_iter=1000,class_weight="balanced")


In [57]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(
    n_estimators=300,
    max_depth=None,
    min_samples_split=5,
    min_samples_leaf=2,
    class_weight="balanced",  # VERY IMPORTANT
    random_state=42,
    n_jobs=-1
)


In [61]:
pipeline.steps[-1]=("model",model)

pipeline.fit(X_train,y_train)

y_prob = pipeline.predict_proba(X_test)[:, 1]


In [62]:
y_pred_custom = []

for p in y_prob:
    if p >=0.3:
        y_pred_custom.append(1)
    else:
        y_pred_custom.append(0)

In [63]:
accuracy = accuracy_score(y_test, y_pred_custom)
precision = precision_score(y_test, y_pred_custom)
recall = recall_score(y_test, y_pred_custom)
f1_value = f1_score(y_test, y_pred_custom)
Confusion_M = confusion_matrix(y_test, y_pred_custom)

print("Accuracy",accuracy)
print("precision",precision)
print("recall",recall)
print("f1_score",f1_value)
print("confusion",Confusion_M)

Accuracy 0.7818791946308725
precision 0.18036286019210246
recall 0.9337016574585635
f1_score 0.3023255813953488
confusion [[2627  768]
 [  12  169]]


# Logistic Regression Evaluation
- Accuracy 0.7818791946308725
- precision 0.18036286019210246
- recall 0.9337016574585635
- f1_score 0.3023255813953488
- confusion [[2627  768]
 [  12  169]]

# Random Forest Evaluation
- Accuracy 0.982662192393736
- precision 0.983739837398374
- recall 0.6685082872928176
- f1_score 0.7960526315789473
- confusion [[3393    2]
 [  60  121]]